### PR TITLE
feat: Allow use of dynamic schemas

### PR DESCRIPTION
**Pull request by Christos-Hadjinikolis**
## Overview

We need to be able to provide validation values in schema definitions that are dynamic. 

Think of a case where you want to say that a `year` field can't have values beyond the current year. You would have to issue a PR to change this at the end of every year so that PRODUCTION code would not break. 

## Key Changes

* Allows schemas to be parsed in a way that replaces dynamic values signified with "[[ DYNAMIIC_VALUE ]]" where `DYNAMIIC_VALUE` is the name of a configuration variable provided as input in an IOConfig object. 

```python
input_config = IOConfig(
    path_to_source_yaml=(os.path.join(RESOURCES, "definitions/input.yaml")),
    env_identifier=ENVIRONMENT,
    dynamic_vars=environment,
)
```
`environment.py `
```

"""A module for configuring all environment variables."""
import os

ENVIRONMENT = "sample"
CLOUD_ENV = "DEV"
RESOURCES = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../resources")
TEST_RESOURCES = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../tests")
S3_YOUR_INPUT_BUCKET = None
S3_YOUR_OUTPUT_BUCKET = None
KAFKA_SERVER = None
KAFKA_TOPIC = None
DB_HOST = None
DB_PORT = None
DB_NAME = None
DB_USER = None
DB_PASS = None
REFERENCE_DATA_STATE_KEY = None
LOWER_THAN_LIMIT = 1000
```

schema file:
```yaml
---
name: bar
columns:
  column_a:
    type: "object"
    validations:
      has_unique_values:
        apply: true
        options: {}
    metrics:
      - Counts
  column_b:
    type: "object"
    validations:
      has_no_null_values:
        apply: true
        options: {}
    metrics:
      - CountsPerLabel
  column_c:
    type: float64
    validations:
      is_greater_than:
        apply: true
        options:
          threshold: 1000
    metrics: []
  column_d:
    type: float64
    validations:
      is_lower_than:
        apply: true
        options:
          threshold: "[[ LOWER_THAN_LIMIT ]]"
    metrics:
      - Min
      - Max
      - Mean
      - Std
      - Variance

```

## Impact

* We can now define schemas with dynamic values. 

## Checklist

The following won't apply in all cases - mark `N/A` next to point if not needed.

- [x] Appropriate unit tests added/updated
- [x] Module, function, class docstrings updated
- [x] Comments added to PR where necessary
- [x] Interface documentation updated
- [x] Semantic commits used for this PR, so that a CHANGELOG can be generated from them (don't delete your local branch! when tagging a new release from master)

## Release-Steps

- [ ] Merge PR
- [ ] Release new tag

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -389,41 +389,24 @@ Let's have a look inside.
 
 ```python
 """A module for configuring all environment variables."""
+"""A module for configuring all environment variables."""
```
Comment (ArthurD1) [inline README.md:392]: This docstring is duplicated.

File: dynamicio/config.py [hunk 2]
```diff
@@ -83,15 +83,16 @@ def with_module(cls, module: ModuleType):
         """
         return type(f"{cls.__name__}_{module.__name__}", (cls,), {"module": module})
 
-    def dyn_str_constructor(self, node: yaml.nodes.ScalarNode) -> str:
-        """Responsible for the switching of one or more "[[ DYNAMIC_VAR ]]" strings with the respective attributes value in a give module.
+    def dyn_str_constructor(self, node: yaml.nodes.ScalarNode, is_schema: bool = False) -> Any:
```
Comment (ArthurD1) [inline dynamicio/config.py:86]: `nit` :why not use 
```str | float```
instead of `Any` for the type hint?

**Review by ArthurD1 — CHANGES_REQUESTED** [review]
Looks good to me.
Just some duplicated text in the README and a suggestion.

**Comment by rugg2** [comment]
Observation in passing: `DYNAMIIC_VALUE` is perhaps a pun but it looks like a typo

File: dynamicio/config.py (hunk 2 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/config.py]: Done

Comment (Christos-Hadjinikolis) [reply dynamicio/config.py]: Actually, mypy takes issue with it. Leaving it as is.

File: README.md [hunk 3]
```diff
@@ -391,39 +391,21 @@ Let's have a look inside.
...
+S3_YOUR_OUTPUT_BUCKET = None
+KAFKA_SERVER = None
+KAFKA_TOPIC = None
+DB_HOST = None
+DB_PORT = None
+DB_NAME = None
+DB_USER = None
+DB_PASS = None
+REFERENCE_DATA_STATE_KEY = None
+LOWER_THAN_LIMIT = 1000         # We will discuss this one later in step 4.
```
Comment (Christos-Hadjinikolis) [inline README.md:408]: Simpler

File: demo/resources/schemas/processed/final_bar.yaml [hunk 4]
```diff
@@ -31,7 +31,7 @@ columns:
       is_lower_than:
         apply: true
         options:
-          threshold: 1000
+          threshold: "[[ LOWER_THAN_LIMIT ]]"
```
Comment (Christos-Hadjinikolis) [inline demo/resources/schemas/processed/final_bar.yaml:34]: Example

File: demo/src/environment.py [hunk 5]
```diff
@@ -1,35 +1,18 @@
...
+S3_YOUR_OUTPUT_BUCKET = None
+KAFKA_SERVER = None
+KAFKA_TOPIC = None
+DB_HOST = None
+DB_PORT = None
+DB_NAME = None
+DB_USER = None
+DB_PASS = None
+REFERENCE_DATA_STATE_KEY = None
+LOWER_THAN_LIMIT = 1000
```
Comment (Christos-Hadjinikolis) [inline demo/src/environment.py:18]: Simpler. 

Of course here once could have something like: 
```python
LOWER_THAN_LIMIT = datetime.datetime.year() + 1
```

File: dynamicio/config.py [hunk 6]
```diff
@@ -52,17 +52,17 @@
             }
         }
 """
-__all__ = ["IOConfig", "SafeDynamicLoader"]
+__all__ = ["IOConfig", "SafeDynamicResourceLoader", "SafeDynamicSchemaLoader"]
```
Comment (Christos-Hadjinikolis) [inline dynamicio/config.py:55]: this was the simplest way to implement this. I could try to parameterise the original loader but it turns out it is more complex than it looks. This keeps things cleaner at least.

File: tests/test_config.py [hunk 7]
```diff
@@ -133,16 +133,42 @@ def test_get_for_config_io_set_for_a_cloud_env_returns_a_cloud_mapping_for_an_po
...
+    @pytest.mark.unit
+    def test_replaces_all_schema_template_instances(self):
+        file_contents = 'abc: "[[ VALUE_A ]]"'
+
+        class MockEnvironmentModule:  # pylint: disable=R0903
+            VALUE_A = 100
+
+        result = yaml.load(io.StringIO(file_contents), SafeDynamicSchemaLoader.with_module(MockEnvironmentModule))
+
+        assert result == {"abc": 100}
```
Comment (Christos-Hadjinikolis) [inline tests/test_config.py:174]: Confirm behaviour

File: dynamicio/config.py (hunk 2 above)
Comment (ArthurD1) [reply dynamicio/config.py]: Oh my bad it must be because you're not on python 3.10, then the right way to write unions is: `Union[str, float]`

**Review by ArthurD1 — APPROVED** [review]
Looks good to me.

File: dynamicio/config.py [hunk 8]
```diff
@@ -104,6 +104,53 @@ def dyn_str_constructor(self, node: yaml.nodes.ScalarNode) -> str:
...
+    def with_module(cls, module: ModuleType):
+        """Creates a dynamic subclass of SafeDynamicLoader with the `data_module` attribute set to `module`.
+
+        Args:
+            module: A global vars module with all the dynamic values defined in it.
+
+        Returns:
+            type
+        """
+        return type(f"{cls.__name__}_{module.__name__}", (cls,), {"module": module})
```
Comment (Christakou) [inline dynamicio/config.py:126]: This line seems interesting, i've not come across this type of syntax, would you mind explaining what is happening here?

File: dynamicio/config.py [hunk 9]
```diff
@@ -104,6 +104,53 @@ def dyn_str_constructor(self, node: yaml.nodes.ScalarNode) -> str:
...
 
 
+class SafeDynamicSchemaLoader(yaml.SafeLoader):
+    """Implements a dynamic yaml loader that parses yaml files and replaces strings that map to [[ DYNAMIC_VAR ]].
+
+    Dynamic variables defined in a provided module object.
+    """
+
+    module = None
+    dynamic_data_matcher = re.compile(r"(.*)(\[\[\s*(\S+)\s*]])(.*)")
```
Comment (Christakou) [inline dynamicio/config.py:114]: I was surprised that we're matching also on anything before and after the  '[[ ' and ' ]]', may I ask why this is the case?

File: dynamicio/config.py (hunk 8 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/config.py:126]: What happens if you want to use something like: `"[[ VALUE_1 ]]/[[ VALUE_2 ]]"'`

File: dynamicio/config.py (hunk 9 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/config.py:114]: What happens if you want to use something like: `"[[ VALUE_1 ]]/[[ VALUE_2 ]]"'`

File: dynamicio/config.py (hunk 8 above)
Comment (RaduG) [reply dynamicio/config.py:126]: Hey Christiano. You might be familiar with the other form of type, which you may have used as `cls = type(some_instance)`, which sets `cls` to the class reference of a`some_instance`'s type.

This other form you see here is `type(cls_name: str, bases: Tuple[Type], members: Mapping[str, Any])` and it returns a new class! So, here, this is creating a new class named `{class name}_{module name}` which is a subclass of the current class and has one attribute `module` set to the object to which `module` points to.

This trick is used to bind a module holding the values that must be replaced to resolve the config to an actual config class.

**Pull request merged**